Implement clone, commit and push of release notes draft to user's fork

puerco opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This is the first (partial) implementation of the release notes PR automation we have been discussing mainly in #1061 . Specifically, this PR adds a new option (--create-draft-pr) to the krel release-notes subcommand that, when enabled, will generate the release notes draft to a user's fork of k/sig-release. The user's fork must be determined using the --draft-org and --draft-repo flags

I also added a new option (-o|--output-dir) to output the release notes in markdown and json to a directory, mainly for testing purposes.

**Which issue(s) this PR fixes**:

Relates to #1061

**Special notes for your reviewer**:
This PR stops short of talking to GitHub and creating the PR. I wanted to touch base first and see if this is going in the right direction.

I added two new functions to pkg/go.git, one to add a remote (Repo.AddRemote(name, owner, repo string)) and another to push to a specific branch (Repo.PushToRemote())  

**Does this PR introduce a user-facing change?**:
```release-note
Add --create-draft-pr --draft-org and --draft-repo flags to automate the creation of the release notes draft
Add --output-dir to output a copy of the release notes.
```
/cc @saschagrunert @JamesLaverack @evillgenius75

justaugustus commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -66,7 +67,10 @@ permissions to your fork of k/sig-release and k-sigs/release-notes.`,
 type releaseNotesOptions struct {
-	tag string
+	tag            string
+	sigReleaseFork string
```
Let's break this up into two fields, so there's no need for string manipulation later.

Use something like:
```go
draftOrg    string
draftRepo   string
```

JamesLaverack replied:
> draftOrg

It might be a user and not an organisation, e.g., `JamesLaverack/sig-release`. Is there a generic term, in GitHub's API, for a "user or organisation"? When dealing with the API do we have to know the difference?

justaugustus replied:
@JamesLaverack -- I think `org` is the correct term from the API perspective.
A user of this tool should understand `--draft-org` to mean "identifier of some user or org", so we should make that clear in the flag's note.

puerco replied:
Ok, the new flags are done in the new revision. The flag uses `org` but in git.go I used `owner` because other parts of the code use that.

justaugustus commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -85,6 +89,30 @@ func init() {
 
+	releaseNotesCmd.PersistentFlags().StringVarP(
+		&releaseNotesOpts.sigReleaseFork,
+		"sigrelease-fork",
```
Same as above; use two options with the flags `draft-org`, `draft-repo`

puerco replied:
Done. New flags are in, old ones gone.

justaugustus commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -85,6 +89,30 @@ func init() {
+		"sigrelease-fork",
+		"",
+		"",
+		"user's fork of k/sig-release repo (eg user/reponame)",
```
Be clear in the usage note that this flag is going to be used to create PRs.

puerco replied:
I've rewritten the help text. I hope it is clearer now.

justaugustus commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -113,12 +141,85 @@ func runReleaseNotes() (err error) {
 	logrus.Infof("Using end tag %v", tag)
 
-	_, err = releaseNotesFrom(start)
+	if releaseNotesOpts.sigReleaseFork == "" && releaseNotesOpts.jsonFile == "" && releaseNotesOpts.mdFile == "" {
```
This should probably be a small function, like: `validatePROptions()`
Add a `BoolVar` flag like `--create-pr` and then you can do something like:
```go
if releaseNoteOpts.createPR {
  prOptsErr := validatePROptions(releaseNotesOpts)
  if prOptsErr != nil {
    return prOptsErr
  }
}
```

puerco replied:
Done. The new flag is called `--create-draft-pr` as we could eventually have another for the JSON version of the release notes.

puerco replied:
Done, `validatePROptions` now exists

justaugustus commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -113,12 +141,85 @@ func runReleaseNotes() (err error) {
 
-	_, err = releaseNotesFrom(start)
+	if releaseNotesOpts.sigReleaseFork == "" && releaseNotesOpts.jsonFile == "" && releaseNotesOpts.mdFile == "" {
+		logrus.Fatal("No output selected. You have to specify at least one of --md-file --json-file or --sigrelease-fork")
```
Surface the error w/ `errors.New()` instead of `logrus.Fatal()`

puerco replied:
Done, all first level errors are now created with errors.New, while I'm wrapping others which are happening further inside.

justaugustus commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -66,7 +67,10 @@ permissions to your fork of k/sig-release and k-sigs/release-notes.`,
+	tag            string
+	sigReleaseFork string
+	mdFile         string
+	jsonFile       string
```
Instead of expecting a filename, how about specifying `outputDir`, with a corresponding `--output-dir` flag?
Then you can default filenames for both the markdown and json output.

puerco replied:
OK, I've implenented this. Maybe we could add a --format flag like in the old release notes tool :thinking:

justaugustus commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -113,12 +141,85 @@ func runReleaseNotes() (err error) {
+	}
+
+	// Clone the user's fork of k/sig-release
+	srrepo, err := git.CloneOrOpenGitHubRepo("", repoparts[0], repoparts[1], true)
```
`srrepo` is an unclear variable. Prefer `repo` or being super explicit with `sigReleaseRepo`.

Additionally, by checking out the user's fork, we're making the assumption that the user's fork is even with the upstream.
We should:
- checkout kubernetes/sig-release
- add the user's fork as a remote
- verify the branch doesn't already exist on the user's fork
- checkout the new branch
- generate the notes
- push

puerco replied:
Done, this sequence now takes place inside createDraftPR()

justaugustus commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -113,12 +141,85 @@ func runReleaseNotes() (err error) {
+
+// createDraftPR pushes the release notes draft to the users fork
+func createDraftPR(tag string, result *releaseNotesResult) error {
+	repoparts := strings.Split(releaseNotesOpts.sigReleaseFork, "/")
```
Actually, we should probably add the `validatePROptions()` check here instead.
`repoparts` becomes unnecessary if the flags get split as suggested above.

puerco replied:
I chose to call `validatePROptions()` before generating the notes. i wanted to avoid running the release notes gatherer (which is very expensive timewise) if the PR was going to fail anyway.

justaugustus commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -113,12 +141,85 @@ func runReleaseNotes() (err error) {
+	targetdir := srrepo.Dir() + fmt.Sprintf("/releases/release-%d.%d", s.Major, s.Minor)
+	logrus.Debugf("Release notes markdown will be written to %s", targetdir)
+
+	// Ugly hack to match repo.Checkout
```
Why is the ugly hack necessary?

puerco replied:
The Commit() function arguments takes a revision and arguments (Checkout(rev string, args ...string)) . I need to pass -b to git but if i send it in the arguments it appends it at the end and it breaks.

justaugustus commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -113,12 +141,85 @@ func runReleaseNotes() (err error) {
+	}
+
+	logrus.Infof("Pushing release notes draft to %s", releaseNotesOpts.sigReleaseFork)
+	err = ioutil.WriteFile(targetdir+"/release-notes-draft.md", []byte(result.markdown), 0644)
```
`targetdir` can probably become the `releaseNotesOpts.outputDir` suggested above.

puerco replied:
Here, do you mean overriding the default location in the repo with --output-dir if defined ?

justaugustus commented on pkg/git/git.go:
```diff
@@ -658,6 +658,17 @@ func (r *Repo) Add(filename string) error {
 }
 
+// CommitAll Commit all changes using git command
+func (r *Repo) CommitAll(msg string) ([]string, error) {
```
Is there a reason `Commit()` doesn't suit our needs here?
If it's the `Author` field, you can push those values into consts instead.

puerco replied:
I resolved this to Add()ing only the release notes files. The reason I was adding a new function was because the Commit functions does not take arguments besides the message.

puerco commented:
Thanks @justaugustus for such detailed guidance, changes are comming right up.

puerco commented:
Due to rebase issues I'm scraping this PR in favour of PR #1110 which includes all your suggestions. I'm still addressing the comments here.

JamesLaverack commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -85,6 +90,38 @@ func init() {
+		"create-draft-pr",
+		"",
+		false,
+		"create the Release Notes draft PR. --draft-org and --draft-repo muste be set along with this option",
```
Typo: "muste" -> "must"
```suggestion
		"create the Release Notes draft PR. --draft-org and --draft-repo must be set along with this option",
```

puerco replied:
Fixed

JamesLaverack commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -113,12 +150,116 @@ func runReleaseNotes() (err error) {
+		if err != nil {
+			return errors.Wrap(err, "Failed to create release notes draft PR")
+		}
+	}
```
If `releaseNotesOpts.createDraftPR` is false, because `--create-draft-pr` was not given, should we log out to say we're skipping it and why? Might help users understand why it's not working if they missed that flag.

puerco replied:
I think if do not specify an output option for the release notes we could print out the help screen. There are and will be more options in addition to --create-draft-pr in the subcommand. For example just writing the release notes to files or generating the json version.

JamesLaverack commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -113,12 +150,116 @@ func runReleaseNotes() (err error) {
+		// Check if --draft-repo is set
+		if releaseNotesOpts.draftRepo == "" {
+			return errors.New("cannot generate Release Notes draft PR without --draft-repo")
+		}
```
It's unfortunate that if you run it without either you'll get an error telling you to give `--draft-org`, then a second time you'll get an error telling you to give `--draft-repo`. As opposed to getting told you need both on the first invocation.

I don't think I have a better idea though. 🤔

puerco replied:
I've now added warning using logrus to advice the user is they are missing either

JamesLaverack commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -113,12 +150,116 @@ func runReleaseNotes() (err error) {
+	}
+
+	// generate the notes
+	targetdir := sigReleaseRepo.Dir() + fmt.Sprintf("/releases/release-%d.%d", s.Major, s.Minor)
```
You probably want to use `filepath.Join` here, which is platform independent.

```suggestion
	targetdir := filepath.Join(sigReleaseRepo.Dir(), "releases", fmt.Sprintf("release-%d.%d", s.Major, s.Minor))
```

puerco replied:
done

JamesLaverack commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -113,12 +150,116 @@ func runReleaseNotes() (err error) {
+	// generate the notes
+	targetdir := sigReleaseRepo.Dir() + fmt.Sprintf("/releases/release-%d.%d", s.Major, s.Minor)
+	logrus.Debugf("Release notes markdown will be written to %s", targetdir)
+	err = ioutil.WriteFile(targetdir+"/release-notes-draft.md", []byte(result.markdown), 0644)
```
As above.
```suggestion
	err = ioutil.WriteFile(filepath.Join(targetdir, "release-notes-draft.md"), []byte(result.markdown), 0644)
```

JamesLaverack replied:
Also probably pull the string `"release-notes-draft.md" `into a constant somewhere, as it's used a few times?

puerco replied:
OK, I've abstracted the filename to a constant:  `DraftFilename`

JamesLaverack commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -113,12 +150,116 @@ func runReleaseNotes() (err error) {
+	}
+
+	// commit the results
+	err = sigReleaseRepo.Add(fmt.Sprintf("releases/release-%d.%d", s.Major, s.Minor) + "/release-notes-draft.md")
```
```suggestion
	err = sigReleaseRepo.Add(filepath.Join("releases", fmt.Sprintf("release-%d.%d", s.Major, s.Minor), "release-notes-draft.md")
```

puerco replied:
Done, all paths are now formed using filepath,Join()

saschagrunert commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -34,6 +36,11 @@ import (
 
+const (
+	// DraftFilename filename for the release notes draft
+	DraftFilename = "release-notes-draft.md"
```
nit: This does not have to be public, right?

puerco replied:
Yes you're right. I thought it could be exposed but there is no need to have it public. Done

saschagrunert commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -85,6 +96,38 @@ func init() {
 		"version tag for the notes",
 	)
 
+	releaseNotesCmd.PersistentFlags().StringVarP(
```
nit: Use `StringVar` if we have no shorthand

puerco replied:
DOne, all instances without short options are now StringVar

saschagrunert commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -85,6 +96,38 @@ func init() {
+		&releaseNotesOpts.outputDir,
+		"output-dir",
+		"o",
+		"",
```
```suggestion
		".",
```

puerco replied:
Done the default is now . and I've ported the --format option for your previous version. If you run it with --create-draft-pr it will not render the release notes in the current directory

saschagrunert commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -113,12 +156,122 @@ func runReleaseNotes() (err error) {
-	_, err = releaseNotesFrom(start)
+	if releaseNotesOpts.createDraftPR {
+		err = validateDraftPROptions()
+		if err != nil {
```
nit
```suggestion
		if err := validateDraftPROptions(); err != nil {
```

puerco replied:
done

saschagrunert commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -113,12 +156,122 @@ func runReleaseNotes() (err error) {
+	// Create RN draft PR
+	if releaseNotesOpts.createDraftPR {
+		err = createDraftPR(tag, result)
+		if err != nil {
```
nit
```suggestion
		if err := createDraftPR(tag, result); err != nil {
```

saschagrunert commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -113,12 +156,122 @@ func runReleaseNotes() (err error) {
+	if releaseNotesOpts.createDraftPR {
+		err = createDraftPR(tag, result)
+		if err != nil {
+			return errors.Wrap(err, "Failed to create release notes draft PR")
```
Nittest of a nit:
```suggestion
			return errors.Wrap(err, "failed to create release notes draft PR")
```
Let's start errors lower case to look homogeneous when showing them to the user.

puerco replied:
I was trying to understand whe or how errors where upper/lower case. Now I now. Done

saschagrunert commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -113,12 +156,122 @@ func runReleaseNotes() (err error) {
+	branchname := "release-notes-draft-" + tag
+
+	// checkout kubernetes/sig-release
+	sigReleaseRepo, err := git.CloneOrOpenGitHubRepo("", "kubernetes", "sig-release", true)
```
We probably should take the repo path from the options. And use `git.DefaultGithubOrg` for the org. For the repo we could define a new variable in the git package, like `DefaultGithubReleaseRepo`

puerco replied:
I;ve added a new option to determine the path: `sigrelease-fork-path` it defaults to os.TempDir(), "k8s-sigrelease". Maybe we can come up with a better name for the flag?

saschagrunert commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -85,6 +96,38 @@ func init() {
+		"draft-org",
+		"",
+		"",
+		"a Github organization or user where the Release Notes PR will be created",
```
We should mention that this is the org of the fork. (what a rhyme)

puerco replied:
Done, the new message reads:
```
a Github organization ownwer of the fork of k/sig-release where the Release Notes Draft PR will be created
```

saschagrunert commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -85,6 +96,38 @@ func init() {
+		"draft-repo",
+		"",
+		"",
+		"the name of a Github repository where the Release Notes PR will be created",
```
We should mention that this is the fork of the repository

puerco replied:
I've altered the help message, now it reads:
```
the name of the fork of k/sig-release, the Release Notes Draft PR will be created from this repository
```

saschagrunert commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -113,12 +156,122 @@ func runReleaseNotes() (err error) {
+	// verify the branch doesn't already exist on the user's fork
+	err = sigReleaseRepo.HasRemoteBranch(branchname)
+	if err == nil {
+		return errors.New(fmt.Sprintf("Remote repo already has a branch named %s", branchname))
```
```suggestion
		return errors.Errorf("remote repo already has a branch named %s", branchname)
```

puerco commented:
Thanks for your comments @saschagrunert I love the attention you give to these kinds of details in your code. I think I've addressed all of your remarks I'll run a new test in a bit (i've run out of my API quota for the next hour).

saschagrunert commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -85,6 +99,49 @@ func init() {
+		&releaseNotesOpts.draftOrg,
+		"draft-org",
+		"",
+		"a Github organization ownwer of the fork of k/sig-release where the Release Notes Draft PR will be created",
```
```suggestion
		"a Github organization owner of the fork of k/sig-release where the Release Notes Draft PR will be created",
```

saschagrunert replied:
See https://github.com/kubernetes/release/pull/1121 (just a nit included there)

saschagrunert commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -85,6 +99,49 @@ func init() {
+	releaseNotesCmd.PersistentFlags().StringVar(
+		&releaseNotesOpts.draftRepo,
+		"draft-repo",
+		"",
```
Lets default that to sig-release too since its the default forks name :)

saschagrunert replied:
See https://github.com/kubernetes/release/pull/1121

saschagrunert commented on cmd/krel/cmd/release_notes.go:
```diff
@@ -113,12 +170,124 @@ func runReleaseNotes() (err error) {
+	}
+
+	// add the user's fork as a remote
+	err = sigReleaseRepo.AddRemote("userfork", releaseNotesOpts.draftOrg, releaseNotesOpts.draftRepo)
```
As a follow-up (I can take care of that as well): If the fork already exists, then we could verify that it matches some pattern, like: `[git@]github.com:$ORG/$REPO`. If this is the case, then we're probably good to proceed

saschagrunert replied:
See https://github.com/kubernetes/release/pull/1122

saschagrunert commented on pkg/git/git.go:
```diff
@@ -747,3 +748,26 @@ func (r *Repo) Rm(force bool, files ...string) error {
 }
+
+// AddRemote adds a new remote to the current working tree
+func (r *Repo) AddRemote(name, owner, repo string) error {
```
As a follow-up (I can take care of that as well): Adding tests for both new API methods.

saschagrunert replied:
See https://github.com/kubernetes/release/pull/1120
